src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -3101,5 +3101,213 @@ internal virtual ImmutableArray<LabelSymbol> Labels
+            {
+                if (member.Kind != SymbolKind.Method)
+                {
+                    candidateMethods.Free();
+
+                    if (warningsOnly)
+                    {
+                        ReportPatternWarning(diagnostics, patternType, member, syntaxExpr, messageID);
+                    }
+                    return null;
+                }
+
+                MethodSymbol method = (MethodSymbol)member;
+
+                // SPEC VIOLATION: The spec says we should apply overload resolution, but Dev10 uses
+                // some custom logic in ExpressionBinder.BindGrpToParams.  The biggest difference
+                // we've found (so far) is that it only considers methods with zero parameters
+                // (i.e. doesn't work with "params" or optional parameters).
+                if (!method.Parameters.Any())
+                {
+                    candidateMethods.Add((MethodSymbol)member);
+                }
+            }
+            MethodSymbol patternMethod = PerformPatternOverloadResolution(patternType, candidateMethods, syntaxExpr, warningsOnly, diagnostics, syntaxTree, messageID);
+
+            candidateMethods.Free();
+
+            return patternMethod;
+        }
+
+        /// <summary>
+        /// The overload resolution portion of FindForEachPatternMethod.
+        /// </summary>
+        private MethodSymbol PerformPatternOverloadResolution
+            (TypeSymbol patternType, ArrayBuilder<MethodSymbol> candidateMethods,
+            SyntaxNode syntaxExpression, bool warningsOnly, DiagnosticBag diagnostics,
+            SyntaxTree syntaxTree, MessageID messageID)
+        {
+            ArrayBuilder<TypeSymbol> typeArguments = ArrayBuilder<TypeSymbol>.GetInstance();
+            AnalyzedArguments arguments = AnalyzedArguments.GetInstance();
+            OverloadResolutionResult<MethodSymbol> overloadResolutionResult = OverloadResolutionResult<MethodSymbol>.GetInstance();
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            // We create a dummy receiver of the invocation so MethodInvocationOverloadResolution knows it was invoked from an instance, not a type
+            var dummyReceiver = new BoundImplicitReceiver(syntaxExpression, patternType);
+            this.OverloadResolution.MethodInvocationOverloadResolution(
+                methods: candidateMethods,
+                typeArguments: typeArguments,
+                receiver: dummyReceiver,
+                arguments: arguments,
+                result: overloadResolutionResult,
+                useSiteDiagnostics: ref useSiteDiagnostics);
+            diagnostics.Add(syntaxExpression, useSiteDiagnostics);
+
+            MethodSymbol result = null;
+
+            if (overloadResolutionResult.Succeeded)
+            {
+                result = overloadResolutionResult.ValidResult.Member;
+
+                if (result.IsStatic || result.DeclaredAccessibility != Accessibility.Public)
```

Comment:
>result.DeclaredAccessibility != Accessibility.Public [](start = 39, length = 52)

Does it really have to be public? Or only accessible here?

Reply:
It looks like the foreach pattern requires public. Let's stick with that for now and I'll add it as a question for LDM in an email.